File: leetcode/c++/11_container_with_most_water.cpp
// https://leetcode.cn/problems/container-with-most-water/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int i = 0, j = height.size() - 1;
        int res = 0;
        while (i < j) {
            res = max(res, (j - i) * min(height[i], height[j]));
            if (height[i] < height[j])
                i++;
            else 
                j--;
        }
        return res;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(1)

File: leetcode/golang/11_container_with_most_water.go
func maxArea(height []int) int {
    i, j := 0, len(height) - 1
    res := 0
    for i < j {
        res = max(res, (j-i) * min(height[i], height[j]))
        if height[i] < height[j] {
            i++
        } else {
            j--
        }
    }
    return res
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}